api/pkg/import/zip_package.go

```diff
@@ -0,0 +1,149 @@
+	}
+
+	srcFile, err := os.Open(zipFile)
+	if err != nil {
+		return fmt.Errorf("could not open file %s: %w", zipFile, err)
+	}
+
+	zipReader, err := zip.NewReader(srcFile, srcInfo.Size())
+	if err != nil {
+		return fmt.Errorf("invalid zip archive: %w", err)
+	}
+
+	extractionDir := strings.TrimSuffix(srcFile.Name(), defaultImportArchiveExtension)
+	err = os.Mkdir(extractionDir, os.ModeDir|os.ModePerm)
+	if err != nil {
+		return fmt.Errorf("error creating temporary extraction directory %s: %w", extractionDir, err)
+	}
+
+	m.extractedDir = extractionDir
+	return extractZipArchive(zipReader, extractionDir, maxSize)
+}
+
+// This function could be part of the manifest interface to abstract the physical location of the files
+// func (m *ZippedPackage) GetResource(resourcePath string) (io.ReadCloser, error) {
+// 	// TODO
+// 	return nil, nil
+// }
+
+// NewPackage creates a new ZippedPackage object ready to be used.
+// The zip file contents will be extracted in a subDirectory with the same name as the file stripped of the .zip
+// extension. During the extraction, zip file uncompressed content is checked not to surpass maxSize.
+// If any error occurs, the temporary folder is cleaned up and (nil, error) will be returned
+func NewPackage(zipFile string, maxSize uint64) (*ZippedPackage, error) {
+	m := new(ZippedPackage)
+	err := m.extract(zipFile, maxSize)
+	if err != nil {
+		m.Close()
+		return nil, fmt.Errorf("error initializing zip import package: %w", err)
+	}
+
+	if _, err = os.Stat(path.Join(m.extractedDir, "manifest.yaml")); err != nil {
+		// error checking for manifest.yaml, cleanup
+		m.Close()
+		return nil, fmt.Errorf("error checking for manifest.yaml: %w", err)
+	}
+
+	return m, nil
+}
+
+func extractZipArchive(reader *zip.Reader, outputDir string, maxSize uint64) error {
+
+	var extractedSize uint64
+	for _, zippedFile := range reader.File {
+		logger.Debugf("Extracting file %+v to %s...", reader, outputDir)
+
+		if zippedFile.FileInfo().IsDir() {
+			fullOutputDirectoryName := path.Join(outputDir, zippedFile.Name)
+			err := os.Mkdir(fullOutputDirectoryName, 0755)
+			if err != nil {
+				return fmt.Errorf("error creating directory %s: %w", fullOutputDirectoryName, err)
+			}
+		} else {
+			err := func() error {
+				if extractedSize+zippedFile.UncompressedSize64 > maxSize {
+					return ErrorUncompressedSizeTooBig
+				}
+
+				dstFileName, written, err := extractZippedFile(zippedFile, outputDir)
+
+				if err != nil {
+					return fmt.Errorf("error extracting %s from archive into %s: %w", zippedFile.Name, dstFileName, err)
+				}
+
+				if written != zippedFile.UncompressedSize64 {
+					logger.Warnf(
+						"Wrong uncompressed size reported for file %s: expected %d, got %d",
+						zippedFile.Name, zippedFile.UncompressedSize64, written,
+					)
+				}
+
+				extractedSize += zippedFile.UncompressedSize64
+				return nil
+			}()
+
+			if err != nil {
+				return err
+			}
```

Comment:
Since we don't do anything fancy with the returned error, I suggest to remove the extra wrapping of a function

Reply:
In this case it's necessary to trigger the defer() calls at the end of each iteration rather than at the end of the function.
The idea is to close  the files as soon as possible (each iteration of the for loop extracts a file) hence the function.
If it looks messy this can be moved to a named private function...